File: .skills/openclaw-skills/skills/nexaiguy/nex-reports/lib/formatter.py
"""
Nex Reports - Output Formatting
MIT-0 License - Copyright 2026 Nex AI (Kevin Blancaflor)
"""

import json
from typing import Any, Dict, List

from .config import REPORT_FOOTER


def format_report_telegram(title: str, module_results: List[Dict[str, Any]]) -> str:
    """
    Format report for Telegram with emojis for status.
    Compact format suitable for mobile viewing.
    """
    emoji_status = {
        "ok": "✓",
        "warning": "⚠",
        "error": "✗",
    }

    lines = [f"📋 {title}"]
    lines.append("")

    for module in module_results:
        status_emoji = emoji_status.get(module.get("status", "ok"), "?")
        title_text = module.get("title", "Module")
        content = module.get("content", "")

        lines.append(f"{status_emoji} {title_text}")
        if content:
            lines.append(f"  {content}")

        items = module.get("items", [])
        if items:
            for item in items[:3]:  # Limit to 3 items per module for Telegram
                lines.append(f"  • {item}")

        lines.append("")

    lines.append(REPORT_FOOTER)

    return "\n".join(lines)


def format_report_markdown(title: str, module_results: List[Dict[str, Any]]) -> str:
    """
    Format report as markdown.
    Full format suitable for email or documents.
    """
    status_emoji = {
        "ok": "✓",
        "warning": "⚠",
        "error": "✗",
    }

    lines = [f"# {title}"]
    lines.append("")
    lines.append(f"*Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}*")
    lines.append("")

    for module in module_results:
        status = module.get("status", "ok")
        status_symbol = status_emoji.get(status, "?")
        module_title = module.get("title", "Module")
        content = module.get("content", "")

        lines.append(f"## {status_symbol} {module_title}")
        lines.append("")

        if content:
            lines.append(f"> {content}")
            lines.append("")

        items = module.get("items", [])
        if items:
            for item in items:
                lines.append(f"- {item}")
            lines.append("")

    lines.append("---")
    lines.append(f"_{REPORT_FOOTER}_")

    return "\n".join(lines)


def format_report_html(title: str, module_results: List[Dict[str, Any]]) -> str:
    """
    Format report as simple HTML.
    """
    from datetime import datetime

    html_lines = [
        "<!DOCTYPE html>",
        "<html>",
        "<head>",
        "<meta charset='UTF-8'>",
        "<meta name='viewport' content='width=device-width, initial-scale=1.0'>",
        "<title>Nex Reports</title>",
        "<style>",
        "body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif; max-width: 900px; margin: 0 auto; padding: 20px; background: #f5f5f5; }",
        "h1 { color: #333; border-bottom: 2px solid #007AFF; padding-bottom: 10px; }",
        ".meta { color: #666; font-size: 12px; margin-bottom: 20px; }",
        ".module { background: white; border-radius: 8px; padding: 15px; margin-bottom: 15px; box-shadow: 0 1px 3px rgba(0,0,0,0.1); }",
        ".module h2 { margin: 0 0 10px 0; font-size: 18px; }",
        ".status-ok { color: #34C759; }",
        ".status-warning { color: #FF9500; }",
        ".status-error { color: #FF3B30; }",
        ".content { color: #555; margin: 10px 0; }",
        ".items { list-style: none; padding: 10px 0; margin: 10px 0; }",
        ".items li { padding: 5px 0; color: #666; }",
        ".footer { text-align: center; color: #999; font-size: 12px; margin-top: 40px; border-top: 1px solid #e0e0e0; padding-top: 20px; }",
        "</style>",
        "</head>",
        "<body>",
        f"<h1>{title}</h1>",
        f"<p class='meta'>Generated: {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}</p>",
    ]

    status_class = {
        "ok": "status-ok",
        "warning": "status-warning",
        "error": "status-error",
    }

    for module in module_results:
        status = module.get("status", "ok")
        module_title = module.get("title", "Module")
        content = module.get("content", "")
        status_cls = status_class.get(status, "status-ok")

        html_lines.append(f"<div class='module'>")
        html_lines.append(f"<h2 class='{status_cls}'>{module_title}</h2>")

        if content:
            html_lines.append(f"<p class='content'>{content}</p>")

        items = module.get("items", [])
        if items:
            html_lines.append("<ul class='items'>")
            for item in items:
                html_lines.append(f"<li>{item}</li>")
            html_lines.append("</ul>")

        html_lines.append("</div>")

    html_lines.append(f"<p class='footer'>{REPORT_FOOTER}</p>")
    html_lines.append("</body>")
    html_lines.append("</html>")

    return "\n".join(html_lines)


def format_report_json(title: str, module_results: List[Dict[str, Any]]) -> str:
    """
    Format report as JSON.
    """
    from datetime import datetime

    report_data = {
        "title": title,
        "generated_at": datetime.now().isoformat(),
        "modules": module_results,
        "footer": REPORT_FOOTER,
    }

    return json.dumps(report_data, indent=2)


def format_report(title: str, results: List[Dict[str, Any]], format_type: str) -> str:
    """
    Dispatcher for format functions.
    format_type: "telegram", "markdown", "html", "json"
    """
    format_map = {
        "telegram": format_report_telegram,
        "markdown": format_report_markdown,
        "html": format_report_html,
        "json": format_report_json,
    }

    formatter = format_map.get(format_type.lower(), format_report_markdown)
    return formatter(title, results)


# Import datetime for HTML formatter
from datetime import datetime

File: .skills/openclaw-skills/skills/homeofe/openclaw-rss-feeds/src/formatter.ts
// @elvatis/openclaw-rss-feeds - HTML digest builder
import type { FeedResult, CveEntry, FirmwareEntry } from './types';

interface DigestMetadata {
  startDate: Date;
  endDate: Date;
  generatedAt?: Date;
}

// Escape HTML entities for safe embedding
function escapeHtml(text: string): string {
  return text
    .replace(/&/g, '&amp;')
    .replace(/</g, '&lt;')
    .replace(/>/g, '&gt;')
    .replace(/"/g, '&quot;')
    .replace(/'/g, '&#39;');
}

// Format a date as "Month YYYY" (e.g. "January 2026")
function formatMonthYear(d: Date): string {
  const months = [
    'January', 'February', 'March', 'April', 'May', 'June',
    'July', 'August', 'September', 'October', 'November', 'December',
  ];
  return `${months[d.getMonth()]} ${d.getFullYear()}`;
}

// Format short date "DD.MM.YYYY"
function formatDate(d: Date): string {
  return d.toLocaleDateString('de-DE', { day: '2-digit', month: '2-digit', year: 'numeric' });
}

// Format CVE description: optionally highlight product names and break on "allows attacker"
function formatCveDescription(desc: string, productHighlightPattern?: string): string {
  let result = desc;

  // Highlight product names if a pattern is configured
  if (productHighlightPattern) {
    try {
      const re = new RegExp(`\\b(${productHighlightPattern})\\b`, 'g');
      result = result.replace(re, '<strong>$1</strong>');
    } catch {
      // Invalid regex - skip highlighting
    }
  }

  // Break before "allows ... attacker"
  result = result.replace(
    /\s+(allows\s+(?:an?\s+)?(?:remote\s+)?attacker)/gi,
    '<br><br>$1'
  );
  return result;
}

function buildSummaryHeader(
  feedResults: FeedResult[],
  metadata: DigestMetadata
): string {
  const totalFirmware = feedResults.reduce((acc, fr) => acc + fr.firmware.length, 0);
  const totalCves = feedResults.reduce((acc, fr) => acc + fr.cves.length, 0);
  const generatedAt = metadata.generatedAt ?? new Date();

  return `
<div style="background:#f8f9fa;border-left:4px solid #1976d2;padding:16px;margin-bottom:24px;font-family:sans-serif;">
  <h3 style="margin:0 0 8px 0;color:#1976d2;">📋 Digest Summary</h3>
  <p style="margin:4px 0;">Period: <strong>${formatDate(metadata.startDate)}</strong> – <strong>${formatDate(metadata.endDate)}</strong></p>
  <p style="margin:4px 0;">Feeds monitored: <strong>${feedResults.length}</strong></p>
  <p style="margin:4px 0;">Firmware updates: <strong>${totalFirmware}</strong></p>
  <p style="margin:4px 0;">CVEs (above threshold): <strong>${totalCves}</strong></p>
  <p style="margin:4px 0;font-size:0.8em;color:#888;">Generated: ${generatedAt.toISOString()}</p>
</div>`;
}

function buildFirmwareSection(firmware: FirmwareEntry[], feedName: string): string {
  if (firmware.length === 0) {
    return `<p style="color:#666;font-style:italic;">No firmware updates for this period.</p>`;
  }

  const rows = firmware
    .map(fw => {
      const typeBg =
        fw.type === 'Major' ? '#e3f2fd' : fw.type === 'Feature' ? '#e8f5e9' : '#fff3e0';
      const typeColor =
        fw.type === 'Major' ? '#1565c0' : fw.type === 'Feature' ? '#2e7d32' : '#e65100';

      const docsCell = fw.docsUrl
        ? `<a href="${escapeHtml(fw.docsUrl)}" target="_blank" style="color:#1976d2;text-decoration:none;">Documentation</a>`
        : `<span style="color:#aaa;">-</span>`;

      return `<tr style="border-bottom:1px solid #eee;">
  <td style="padding:8px;vertical-align:middle;"><strong>${escapeHtml(fw.product)}</strong></td>
  <td style="padding:8px;vertical-align:middle;font-family:monospace;">${escapeHtml(fw.version)}</td>
  <td style="padding:8px;vertical-align:middle;text-align:center;">
    <span style="background:${typeBg};color:${typeColor};padding:2px 8px;border-radius:4px;font-size:0.85em;font-weight:bold;">${fw.type}</span>
  </td>
  <td style="padding:8px;vertical-align:middle;">${docsCell}</td>
</tr>`;
    })
    .join('\n');

  return `<table cellpadding="0" cellspacing="0" style="width:100%;border-collapse:collapse;font-size:0.9em;margin-bottom:8px;">
<thead style="background:#f5f5f5;">
  <tr>
    <th style="padding:8px;text-align:left;border-bottom:2px solid #ddd;">Product</th>
    <th style="padding:8px;text-align:left;border-bottom:2px solid #ddd;">Version</th>
    <th style="padding:8px;text-align:center;border-bottom:2px solid #ddd;">Type</th>
    <th style="padding:8px;text-align:left;border-bottom:2px solid #ddd;">Documentation</th>
  </tr>
</thead>
<tbody>
${rows}
</tbody>
</table>`;
}

function buildCveSection(cves: CveEntry[], cvssThreshold: number, productHighlightPattern?: string): string {
  if (cves.length === 0) {
    return `<p style="color:#666;font-style:italic;">No CVEs above the CVSS threshold (${cvssThreshold}) for this period.</p>`;
  }

  const rows = cves
    .map(cve => {
      const score = cve.score;
      const scoreColor = score >= 9.0 ? '#b71c1c' : score >= 7.0 ? '#d32f2f' : '#e65100';
      const scoreBg = score >= 9.0 ? '#ffcdd2' : score >= 7.0 ? '#ffebee' : '#fff3e0';

      return `<tr style="border-bottom:1px solid #ffebee;">
  <td style="padding:8px;vertical-align:top;white-space:nowrap;">
    <a href="${escapeHtml(cve.url)}" target="_blank" style="color:#d32f2f;font-weight:bold;text-decoration:none;">${escapeHtml(cve.id)}</a>
  </td>
  <td style="padding:8px;vertical-align:top;text-align:center;">
    <span style="background:${scoreBg};color:${scoreColor};padding:2px 8px;border-radius:4px;font-weight:bold;">${score.toFixed(1)}</span>
  </td>
  <td style="padding:8px;vertical-align:top;font-size:0.88em;line-height:1.5;">${formatCveDescription(escapeHtml(cve.description), productHighlightPattern)}</td>
</tr>`;
    })
    .join('\n');

  return `<table cellpadding="0" cellspacing="0" style="width:100%;border-collapse:collapse;font-size:0.9em;margin-bottom:8px;">
<thead style="background:#ffebee;">
  <tr>
    <th style="padding:8px;text-align:left;border-bottom:2px solid #ffcdd2;width:15%;">CVE ID</th>
    <th style="padding:8px;text-align:center;border-bottom:2px solid #ffcdd2;width:10%;">Score</th>
    <th style="padding:8px;text-align:left;border-bottom:2px solid #ffcdd2;">Description</th>
  </tr>
</thead>
<tbody>
${rows}
</tbody>
</table>`;
}

function buildFeedSection(feedResult: FeedResult, cvssThreshold: number): string {
  const hasFirmware = feedResult.firmware.length > 0;
  const hasCves = feedResult.cves.length > 0 && feedResult.enrichCve !== false;

  let section = `<div style="margin-bottom:40px;">
<h2 style="color:#333;border-bottom:2px solid #1976d2;padding-bottom:8px;">📡 ${escapeHtml(feedResult.feedName)}</h2>`;

  if (feedResult.error) {
    section += `<div style="background:#fff3e0;border-left:4px solid #ff9800;padding:12px;margin-bottom:16px;">
  <strong>⚠️ Feed Error:</strong> ${escapeHtml(feedResult.error)}
</div>`;
  }

  // Firmware section
  section += `<h3 style="color:#555;">📦 Firmware Updates</h3>`;
  section += buildFirmwareSection(feedResult.firmware, feedResult.feedName);

  // Items without version (general news items)
  const nonFirmwareItems = feedResult.items.filter(item => !item.version);
  if (nonFirmwareItems.length > 0) {
    section += `<h3 style="color:#555;">📰 Feed Items</h3><ul style="font-size:0.9em;line-height:1.8;">`;
    for (const item of nonFirmwareItems) {
      section += `<li><a href="${escapeHtml(item.link)}" target="_blank" style="color:#1976d2;">${escapeHtml(item.title)}</a>`;
      section += ` <span style="color:#999;font-size:0.85em;">(${formatDate(item.pubDate)})</span></li>`;
    }
    section += `</ul>`;
  }

  if (!hasFirmware && nonFirmwareItems.length === 0 && !feedResult.error) {
    section += `<p style="color:#666;font-style:italic;">No items found for this feed in the selected period.</p>`;
  }

  // CVE section (only if enrichCve is enabled for this feed)
  if (hasCves || feedResult.cveError) {
    section += `<h3 style="color:#555;">⚠️ Security Advisories (CVEs)</h3>`;
    if (feedResult.cveError) {
      section += `<div style="background:#fff3e0;border-left:4px solid #ff9800;padding:12px;margin-bottom:16px;">
  <strong>⚠️ CVE Fetch Warning:</strong> ${escapeHtml(feedResult.cveError)}
</div>`;
    }
    section += buildCveSection(feedResult.cves, cvssThreshold, feedResult.productHighlightPattern);
  }

  section += `</div>`;
  return section;
}

export function formatDigest(
  feedResults: FeedResult[],
  metadata: DigestMetadata,
  cvssThreshold = 6.5
): string {
  const monthYear = formatMonthYear(metadata.startDate);

  // Collect all CVEs across all feeds for the global summary section
  const allCves = feedResults.flatMap(fr => fr.cves);
  allCves.sort((a, b) => b.score - a.score);

  let html = `<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta name="viewport" content="width=device-width,initial-scale=1.0">
  <style>
    body { font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', sans-serif; line-height:1.6; color:#333; max-width:960px; margin:0 auto; padding:20px; }
    a { color:#1976d2; }
    h1 { color:#1a1a1a; }
  </style>
</head>
<body>
<h1>🛡️ Security &amp; Firmware Digest | ${escapeHtml(monthYear)}</h1>
`;

  html += buildSummaryHeader(feedResults, metadata);

  // Per-feed sections
  for (const feedResult of feedResults) {
    html += buildFeedSection(feedResult, cvssThreshold);
  }

  // Global CVE summary (if multiple feeds had CVEs)
  if (feedResults.length > 1 && allCves.length > 0) {
    html += `<div style="margin-top:40px;padding-top:20px;border-top:2px solid #eee;">
<h2 style="color:#333;">🔐 All CVEs - Combined View</h2>
<p style="color:#666;font-size:0.9em;">All CVEs from all feeds above the CVSS threshold (${cvssThreshold}), sorted by severity:</p>
${buildCveSection(allCves, cvssThreshold)}
</div>`;
  }

  html += `\n</body>\n</html>`;
  return html;
}

// Also export a plain-body variant (without full HTML document wrapper) for Ghost
export function formatDigestBody(
  feedResults: FeedResult[],
  metadata: DigestMetadata,
  cvssThreshold = 6.5
): string {
  const monthYear = formatMonthYear(metadata.startDate);
  const allCves = feedResults.flatMap(fr => fr.cves);
  allCves.sort((a, b) => b.score - a.score);

  let html = `<h1>🛡️ Security &amp; Firmware Digest | ${escapeHtml(monthYear)}</h1>\n`;
  html += buildSummaryHeader(feedResults, metadata);

  for (const feedResult of feedResults) {
    html += buildFeedSection(feedResult, cvssThreshold);
  }

  if (feedResults.length > 1 && allCves.length > 0) {
    html += `<div style="margin-top:40px;padding-top:20px;border-top:2px solid #eee;">
<h2 style="color:#333;">🔐 All CVEs - Combined View</h2>
${buildCveSection(allCves, cvssThreshold)}
</div>`;
  }

  return html;
}

// Re-export type for index.ts
export type { DigestMetadata };
